### PR TITLE
Pre-assigning participants and groups to breakout rooms

### LINKED ISSUE #110: Support Breakout Rooms
(not checked the Zoom API if this is supported but...)
It would be *really* nice if you could specify a Moodle grouping that would be used to (pre-)configure Zoom breakout rooms, with a breakout room per group in the grouping.

### PR DESCRIPTION
**Pull request by annouarf**
Hello Jonathan,

My name is Annouar, I am currently working for a university in Canada, and we had to make a few changes to your plugin.
We therefore wanted to share these changes with you, in case they might interest you.

If you have any questions please let us to know.

Have a great day.

### CODE REVIEW

**Comment by jrchamp** [comment]
Thanks @annouarf! This may address #110 and would supersede #115

It looks like there may be some merge conflicts with recent changes to the plugin. Is that something that you or the university would be willing to resolve? It's okay to say no.

**Comment by annouarf** [comment]
Hello Jonathan,

Sorry for the misunderstanding, but we've updated the pull request, and we've deleted .orig files. Many thanks for your patience!

**Comment by annouarf** [comment]
Hello Jonathan,

Sorry for the misunderstanding, but we updated the pull request by rebasing to the new 4.5.0 version and of course fixing the conflicts.
Thank you very much for your patience !

Have a good day.

**Comment by jrchamp** [comment]
Thank you! I'll take a closer look at the code soon and I've approved the automated tests.

**Review by jrchamp — COMMENTED** [review]
There's a lot of little comments here, but that's because I tried to comment on as many things as I could to help minimize the number of changes in the next round of review. Most of the items are from the GitHub Action output.

Notably, I'm not really requesting a change to the functionality, but more improvements to make it pass the automated testing, compatible with multiple database backends, maintainability and accessibility.

You've already done so much work, so if I'm asking too much, please let me know.

**Comment by annouarf** [comment]
Hello Jonathan,

we updated the pull request by resolving almost all your comments.
Thank you very much for your patience !

Have a good day.

File: db/upgrade.php [hunk 1]
```diff
@@ -757,5 +757,59 @@ function xmldb_zoom_upgrade($oldversion) {
...
+        // Adding keys to table zoom_rooms_groups.
+        $table->add_key('primary', XMLDB_KEY_PRIMARY, ['id']);
+        $table->add_key('fk_breakoutroomid', XMLDB_KEY_FOREIGN, array('breakoutroomid'),
+            'zoom_meeting_breakout_rooms', array('id'));
+
+        // Conditionally launch create table for customfield_category.
+        if (!$dbman->table_exists($table)) {
+            $dbman->create_table($table);
+        }
+    }
```
Comment (jrchamp) [inline db/upgrade.php:812]: The `upgrade_mod_savepoint()` is needed to track the upgrade version.

File: templates/breakoutrooms/room/data.mustache [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    *courseparticipants
+      * participantid 
+      * participantemail 
+      * selected 
+    *coursegroups
+      * groupid 
+      * groupname 
+      * selected 
+ }}
+<label for="room-name-{{roomid}}"">Room name</label>
```
Comment (jrchamp) [inline templates/breakoutrooms/room/data.mustache:32]:
```suggestion
<label for="room-name-{{roomid}}">Room name</label>
```

**Review by jrchamp — COMMENTED** [review]
It looks like there's a merge conflict, so I can't run the automated tests against the updated code. I noticed a couple items, but I didn't do a full review because it's better to make sure I'm not duplicating the work that the automated tests would do.

**Comment by annouarf** [comment]
Hello Jonathan,

we updated the pull request by rebasing to the new 4.5.1 version and fixing all conflicts.
Thank you very much for your patience !

Have a good day.

**Comment by annouarf** [comment]
we just pushed some missed fixes.

**Review by jrchamp — COMMENTED** [review]
This is looking very close, the automated tests are almost passing. I added some suggestions that should fix the mustache context issues. For the grunt compilation warnings, you will probably need to run the `grunt amd` task on Moodle's updated MOODLE_311_STABLE branch. If you have any questions, please let me know.

**Comment by annouarf** [comment]
It should be ok now.

**Review by jrchamp — COMMENTED** [review]
Almost ready for testing. After the duplicate context is removed from rooms.mustache, the only other automated error is that `amd/build/form.min.js` and `amd/build/form.min.js.map` are "stale" according to grunt/Moodle 3.11.

I would help fix the problems, but I do not have write access to your branch. Unfortunately, it's the `main` branch, so I wouldn't really want write access. I don't think it is possible to switch a pull request's branch, so we might be out of luck.

**Comment by levemar** [comment]
Hi @jrchamp!

I work with Annouar (@annouarf) at the University of Montreal. I just gave you write access to our repository. I know you said that you "wouldn't really want write access" on the main branch, but don't worry, this repo is only to share our developments with you. We have an other git system for our internal version. 

I think it will be easier this way, since we don't know how to get rid of the "stale" status with grunt. Grunt does not show this error on our local environments, so we can't really reproduce it or correct it...

Your help is greatly appreciated! 
Marie-Eve

**Comment by sgrandh3** [comment]
Found a couple of issues.
1.The same student or a group can be added to multiple rooms.
2. When 2 rooms have identical names, only the room with most recent change is saved. 

Did not verify the behavior of students allocation to the rooms in the above 2 scenarios

**Review by jrchamp — COMMENTED** [review]
@armccoy @smbader Thanks again for offering to help review this. I added a few questions and suggestions.

@sgrandh3 raised some good questions about the ability to assign the same users and groups to multiple rooms. I don't know of a clean way to adjust the JavaScript to prevent that situation, so I'm tempted to document it instead of trying to fix it in a way that might instead cause more problems. If anyone has a better idea, it could even be merged in a future pull request.

File: amd/src/form.js [hunk 3]
```diff
@@ -153,6 +156,258 @@ define(['jquery'], function($) {
...
+                // Changing participants and groups select options values.
+                $("#" + roomParticipantsSelectId + " option").each(function() {
+                    var thisValue = $(this).attr('value').replace('-', '');
+                    $(this).attr('value', thisValue);
+                });
+
+                $("#" + roomGroupsSelectId + " option").each(function() {
+                    var thisValue = $(this).attr('value').replace('-', '');
+                    $(this).attr('value', thisValue);
+                });
```
[the remark this answers is not in the document]
Comment (annouarf) [reply amd/src/form.js:343]: No, It is not needed

**Comment by annouarf** [comment]
Hello Jonathan,

for this issue : 
**The same student or a group can be added to multiple rooms.**

I think that when we add a participant X to a room, we must remove it from the drop-down lists of the other rooms.
and vice versa when we remove a participant X from a room, we must add it to the drop-down lists of the other rooms.

But the question that arises is how to catch the events (select/unselect) of the autocomplete element ?

**Comment by jrchamp** [comment]
Hello @annouarf,

I agree that would be a good approach, but I also agree that catching the events may be exceedingly complicated. For now, I'm content with the current behavior.

@smbader has been working on the last few issues regarding string translation and we should hopefully have something we can merge soon.

**Review by jrchamp — APPROVED** [review]
Looks like we're ready to merge.

The most recent commit addressed the "buttons default to behaving as submit buttons" issue that was causing rooms to be deleted when pressing `Enter` while focused on a form input.

**Pull request merged**